keps/77-dynamically-sized-jobs/README.md

```diff
@@ -7,168 +7,290 @@
+At most two active workload slices can exist for any single Job in Kueue: the current admitted slice and a newly created slice during a scaling operation. 
+This ensures a controlled transition between slices while maintaining consistency and avoiding resource contention and race conditions.
+
+To facilitate precise resource scheduling and effective workload preemption, every new job configured with slice enablement will be appropriately annotated using `EnabledAnnotationKey/Value` mentioned above.
+Furthermore, each individual workload slice will record the identity of any preempting slice it encounters.
 
 ```golang
-type GenericJob interface {
-  ...
-  ...
-  ResizeJob(wl *kueue.Workload) error
-}
+
+const (
+	// WorkloadPreemptibleSliceNameKey is the annotation key used to capture an "old" workload slice
+	// that will be preempted by the "new" workload slice, i.e., this annotates a "new" workload slice.
+	// This annotation is alpha-level.
+	WorkloadPreemptibleSliceNameKey = "kueue.x-k8s.io/workload-preemptible-slice"
+)
 ```
-Jobs implementing the ResizeJob method will create a Workload Slice for every new replica of a job. 
 
-On scale down to M we will change the original Workload's resources and then on scale up to N we will create N-M WorkloadSlice objects that go through admission and scheduling gate checks, i.e. the Workload can "lose" the original (first time admission) quota when scaling down before scaling up.
+##### Naming
+Currently, Kueue uses a deterministic Workload naming strategy by generating consistent names for the same Job instance via `GetWorkloadNameForOwnerWithGVK`.
+As a result, Workload names are not unique across different revisions of the same owner object (e.g., a Job), which makes this approach unsuitable for representing Workload slices.
+
+To address this, a new naming strategy will be introduced specifically for Workload slices, incorporating the owner object's generation value to ensure name uniqueness across revisions.
+For example, `bachv1/Job`: 
+```yaml
+annotations:
+  kueue.x-k8s.io/elastic-job: "true"
+creationTimestamp: "2025-06-12T21:32:41Z"
+generation: 1
+labels:
+  kueue.x-k8s.io/queue-name: demo
+name: demo-slice
+namespace: demo
+resourceVersion: "190175"
+uid: 78d1e61a-c399-4c09-800f-a0bbfb5fe71e
+```
 
-### Pod Scheduling Gates
+will result in a Workload name similar to: `job-demo-slice-75e0c`
+Whereas, changing `generation: 2`: will result in a different name similar to: `job-demo-slice-e7080`
 
-Inside the job's webhook, implement schedulingGate injection for pods on creation time.
-The Pods will be ungated following a similar behavior as to how a job is suspended and then unsuspended in the when admitted.
-When the job scales up, the new pods will be gated due to the schedulingGates injection in the webhook.
+#### Scheduling and Preemption
```

Comment:
What if the flavor assigner assigns a different flavor to a new Workload from an old Workload?

Will you implement any mechanism to assign the same flavor to a new Workload as an old Workload?
Will the flavor assigner filter out the different flavors from an old Workload?

Reply:
This is actually captured in the **Design Details** section:

> *Each WorkloadSlice inherits the resource flavor assignment of the original admitted slice, enforcing flavor consistency across scale-ups.*

> Will you implement any mechanism to assign the same flavor to a new Workload as an old Workload?

Yes, that’s already a standing requirement.

> Will the flavor assigner filter out the different flavors from an old Workload?

I’m not sure I fully understand the question, could you clarify what you mean by “filter out”?

Reply:
> The assigner just retries the same flavor with increased count, right?

Yes, that is correct.